aea/aea_builder.py

```diff
@@ -0,0 +1,164 @@
+
+PackageId = Tuple[AgentComponentType, PublicId]
+
+
+class _Package:
+
+    def __init__(self, package_type, package_configuration):
+        """"""
+
+        self._package_type = package_type  # one of protocol, skill, connection etc.
+        self._package_config = package_configuration
+        self._modules = {}
+
+    @classmethod
+    def load(self, package_type) -> "_Package":
+        """Load package from dir"""
+        # read config file
+        # load modules
+
+
+class AEABuilder:
+    """This class helps to build an AEA."""
+
+    def __init__(self):
+        self._wallet = Wallet({})  # TODO make Wallet mutable (add/remove_private_key)
+        self._ledger_apis = LedgerApis({}, "")  # TODO make it mutable (add/remove api)
+
+        self._resources = Resources()
+        self._connections = []
+
+        # identity
+        self._name = None
+        self._addresses = {}  # type: Dict[str, Address]
+        self._default_key = None  # set by the user, or instantiate a default one.
+
+        # from package id to packages it depends on.
+        # we should detect loops at "add" time, by inspecting the component config file.
+        self._dependency_graph = {}  # type: Dict[PackageId, Set[_Package]]
+
+        # add default protocol
+        # self.add_protocol()
+        # add error skill
+        # self.add_skill()
+
+    def set_name(self, name: str):
+        self._name = name
+
+    def add_address(self, identifier: str, address: Address):
+        self._addresses[identifier] = address
+
+    def remove_address(self, identifier: str):
+        self._addresses.pop(identifier, None)
+
+    def add_package(self, package_type, directory):
+        """Load a package, given its type and the directory."""
+        # load config, check fingerprint, handle errors
+        # from config, check package dependencies
+        # visit dependency graph, load modules in sys.modules when going backward
+        # try to load the package (e.g. the same we do for Skill.from_dir, but for any type
+        # _Package.load(package_type, directory)
+        # remove all the modules from sys.modules
+        # add the loaded modules in self._agent_modules
+        # update dependency graph
+        # register new package in resources
+
+    def remove_package(self, package_type, directory):
+        """Remove a package"""
+        # check dependency graph for pending packages
+        # remove package
+        # remove modules of package in internal indexes.
+
+    def add_protocol(self, path):
+        """Add a protocol to the agent."""
+        # call add_package
+
+    def remove_protocol(self):
+        """Remove protocol"""
+        # call remove_package
+
+    # def add_connection(self, path):
+    # def remove_connection(self):
+    # def add_skill(self, path):
+    # def remove_skill(self):
+
+    def build(self) -> AEA:
+        """Get the AEA."""
+        aea = AEA(
+            Identity(self._name, addresses=self._addresses),
```

Comment:
would instantiate this as part of builder. Note. addresses come straight out of the wallet